File: go/parser/kind9321.go
package parser

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/candypoets/nutscash/types"
	"github.com/elnosh/gonuts/cashu"
	"github.com/nbd-wtf/go-nostr"
)

// Kind9321Parsed represents parsed data from a kind 9321 event (nutzap)
type Kind9321Parsed struct {
	Amount       int                `json:"amount" msgpack:"amount"`
	Recipient    string             `json:"recipient" msgpack:"recipient"`
	EventID      string             `json:"eventId,omitempty" msgpack:"eventId,omitempty"`       // event being zapped if any
	MintURL      string             `json:"mintUrl" msgpack:"mintUrl"`                           // mint for the proofs
	Redeemed     bool               `json:"redeemed" msgpack:"redeemed"`                         // Whether the zap has been redeemed
	Proofs       []types.ProofUnion `json:"proofs" msgpack:"proofs"`                             // Using ProofUnion to handle both Proof and ProofV4
	Comment      string             `json:"comment,omitempty" msgpack:"comment,omitempty"`       // Optional comment from the sender
	IsP2PKLocked bool               `json:"isP2PKLocked" msgpack:"isP2PKLocked"`                 // Whether the proofs are properly P2PK-locked
	P2PKPubkey   string             `json:"p2pkPubkey,omitempty" msgpack:"p2pkPubkey,omitempty"` // The P2PK pubkey detected in the proofs
}

// ParseKind9321 parses a kind 9321 event (nutzap)
func (p *Parser) ParseKind9321(event nostr.Event) (*Kind9321Parsed, *[]types.Request, error) {
	// Create the necessary request objects
	var requests = []types.Request{}
	if event.Kind != 9321 {
		return nil, nil, fmt.Errorf("event is not kind 9321")
	}

	// get the sender profile for this zap
	requests = append(requests, types.Request{
		Kinds:      []int{0},
		Authors:    []string{event.PubKey},
		CacheFirst: true,
		Relays:     p.GetRelays(event),
	})

	// Extract required tags
	var proofTags [][]string
	var mintTag, recipientTag, eventTag []string

	for _, tag := range event.Tags {
		if len(tag) >= 2 {
			if tag[0] == "proof" {
				proofTags = append(proofTags, tag)
			} else if tag[0] == "u" && mintTag == nil {
				mintTag = tag
			} else if tag[0] == "p" && recipientTag == nil {
				recipientTag = tag
				// Get recipient profile
				requests = append(requests, types.Request{
					Kinds:      []int{0},
					Authors:    []string{recipientTag[1]},
					CacheFirst: true,
					Relays:     p.GetRelays(event),
				})
				// try to find receipt event from the recipient
				var relays []string
				walletRelays, exist := p.DB.QueryEvent(nostr.Filter{Kinds: []int{10019}, Authors: []string{event.PubKey}})
				walletRelaysParsed, ok := walletRelays.Parsed.(*Kind10019Parsed)
				if exist && ok {
					relays = append(relays, walletRelaysParsed.ReadRelays...)
				}
				requests = append(requests, types.Request{
					Kinds: []int{7376},
					Tags: map[string][]string{
						"#e": {event.ID},
					},
					Authors:    []string{recipientTag[1]},
					Limit:      1,
					CacheFirst: true,
					Relays:     relays,
				})

			} else if tag[0] == "e" && eventTag == nil {
				eventTag = tag
				requests = append(requests, types.Request{
					Kinds:      []int{1},
					IDs:        []string{eventTag[1]},
					CacheFirst: true,
					Relays:     p.GetRelays(event),
				})
			}
		}
	}

	// Validate essential tags are present
	if len(proofTags) == 0 || mintTag == nil || recipientTag == nil {
		return nil, nil, fmt.Errorf("missing required tags")
	}

	// Parse nutzap information
	total := 0
	proofs := make([]types.ProofUnion, 0, len(proofTags))
	isP2PKLocked := false
	var p2pkPubkey string

	for _, proofTag := range proofTags {
		// Try to extract proof from the tag
		var proofUnion types.ProofUnion
		if err := json.Unmarshal([]byte(proofTag[1]), &proofUnion); err != nil {
			// Just log the error and continue
			fmt.Printf("Failed to parse proof JSON: %v\n", err)
			continue
		}

		// Get amount based on proof type
		var amount int
		var secretStr string

		// Get the actual proof data (could be Proof or ProofV4)
		actualProof := proofUnion.GetProof() // Assume this method returns the underlying proof interface{}

		switch p := actualProof.(type) {
		case cashu.Proof: // Handle V3 Proof type
			amount = int(p.Amount) // Cast uint64 to int
			secretStr = p.Secret
		case cashu.ProofV4: // Handle V4 Proof type
			amount = int(p.Amount) // Cast uint64 to int
			secretStr = p.Secret   // Assuming V4 also has Amount and Secret fields with these names
		case map[string]any: // Fallback if GetProof returns a map (e.g., if ProofUnion wasn't fully hydrated)
			fmt.Println("GetProof returned map[string]interface{}, attempting map access")
			if amt, exists := p["amount"]; exists {
				switch v := amt.(type) {
				case float64:
					amount = int(v)
				case int:
					amount = v
				case json.Number:
					if intVal, err := v.Int64(); err == nil {
						amount = int(intVal)
					} else {
						fmt.Printf("Fallback: Failed to convert json.Number amount: %v\n", err)
					}
				default:
					fmt.Printf("Fallback: Unexpected type for amount: %T\n", v)
				}
			} else {
				fmt.Printf("Fallback: Proof map does not contain 'amount' key.\n")
			}
			// Attempt to get secret from map as well
			if secret, exists := p["secret"]; exists {
				if s, ok := secret.(string); ok {
					secretStr = s
				}
			}
		default:
			// Handle unexpected type from GetProof()
			fmt.Printf("Unexpected proof type from GetProof(): %T\n", p)
		}

		total += amount
		proofs = append(proofs, proofUnion)

		// Check if this proof is P2PK-locked using the extracted secretStr
		if secretStr != "" && strings.Contains(secretStr, "P2PK") {
			isP2PKLocked = true // Mark zap as P2PK if at least one proof is

			// Try to extract the pubkey only if we haven't found one yet for the whole zap
			if p2pkPubkey == "" {
				var secretData []interface{}
				// Use json.Unmarshal on the secret string
				if err := json.Unmarshal([]byte(secretStr), &secretData); err == nil {
					// Expected format: ["P2PK", {"data": "<pubkey>", ...}]
					if len(secretData) >= 2 {
						if label, ok := secretData[0].(string); ok && label == "P2PK" {
							if dataMap, ok := secretData[1].(map[string]interface{}); ok {
								if dataStr, ok := dataMap["data"].(string); ok {
									p2pkPubkey = dataStr // Store the first valid pubkey found
								}
							}
						}
					}
				} else {
					fmt.Printf("Failed to unmarshal P2PK secret string '%s': %v\n", secretStr, err)
				}
			}
		}
	}

	// Create the parsed result
	result := &Kind9321Parsed{
		Amount:       total,
		Recipient:    recipientTag[1],
		MintURL:      mintTag[1],
		Proofs:       proofs,
		Redeemed:     false, // Default to not redeemed, will check later
		Comment:      event.Content,
		IsP2PKLocked: isP2PKLocked,
		P2PKPubkey:   p2pkPubkey,
	}

	// Add eventId if present
	if eventTag != nil {
		result.EventID = eventTag[1]
	}

	return result, &requests, nil
}

// PrepareKind9321 prepares a kind 9321 event for publishing
func (p *Parser) PrepareKind9321(event *nostr.Event) error {
	if event.Kind != 9321 {
		return fmt.Errorf("event is not kind 9321")
	}

	// Validate required tags
	var hasProof, hasMint, hasRecipient bool

	for _, tag := range event.Tags {
		if len(tag) >= 2 {
			if tag[0] == "proof" {
				hasProof = true
			} else if tag[0] == "u" {
				hasMint = true
			} else if tag[0] == "p" {
				hasRecipient = true
			}
		}
	}

	if !hasProof {
		return fmt.Errorf("kind 9321 must include at least one proof tag")
	}

	if !hasMint {
		return fmt.Errorf("kind 9321 must include a u tag with mint URL")
	}

	if !hasRecipient {
		return fmt.Errorf("kind 9321 must include a p tag with recipient")
	}

	// Sign the event
	return p.Signer.SignEvent(event)
}

File: src/model/nostr-worker/src/parser/kind9321.rs
use crate::parser::Parser;
use crate::types::network::Request;
use crate::utils::request_deduplication::RequestDeduplicator;
use anyhow::{anyhow, Result};
use nostr::{Event, UnsignedEvent};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProofUnion {
    // Placeholder for proof types - would need actual cashu types
    pub data: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Kind9321Parsed {
    pub amount: i32,
    pub recipient: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "eventId")]
    pub event_id: Option<String>,
    #[serde(rename = "mintUrl")]
    pub mint_url: String,
    pub redeemed: bool,
    pub proofs: Vec<ProofUnion>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
    #[serde(rename = "isP2PKLocked")]
    pub is_p2pk_locked: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "p2pkPubkey")]
    pub p2pk_pubkey: Option<String>,
}

impl Parser {
    pub fn parse_kind_9321(&self, event: &Event) -> Result<(Kind9321Parsed, Option<Vec<Request>>)> {
        if event.kind.as_u64() != 9321 {
            return Err(anyhow!("event is not kind 9321"));
        }

        let mut requests = Vec::new();

        // Get the sender profile for this nutzap
        requests.push(Request {
            authors: vec![event.pubkey.to_hex()],
            kinds: vec![0],
            relays: self
                .database
                .find_relay_candidates(0, &event.pubkey.to_hex(), &false),
            close_on_eose: true,
            cache_first: true,
            ..Default::default()
        });

        // Extract required tags
        let mut proof_tags = Vec::new();
        let mut mint_tag: Option<Vec<String>> = None;
        let mut recipient_tag: Option<Vec<String>> = None;
        let mut event_tag: Option<Vec<String>> = None;

        for tag in &event.tags {
            let tag_vec = tag.as_vec();
            if tag_vec.len() >= 2 {
                match tag_vec[0].as_str() {
                    "proof" => proof_tags.push(tag_vec.to_vec()),
                    "u" => {
                        if mint_tag.is_none() {
                            mint_tag = Some(tag_vec.to_vec());
                        }
                    }
                    "p" => {
                        if recipient_tag.is_none() {
                            recipient_tag = Some(tag_vec.to_vec());
                            // Get recipient profile
                            requests.push(Request {
                                authors: vec![tag_vec[1].clone()],
                                kinds: vec![0],
                                relays: self.database.find_relay_candidates(0, &tag_vec[1], &false),
                                cache_first: true,
                                ..Default::default()
                            });

                            // Check for spending history events
                            let mut spending_tags = std::collections::HashMap::new();
                            spending_tags.insert("#e".to_string(), vec![event.id.to_hex()]);

                            requests.push(Request {
                                authors: vec![tag_vec[1].clone()],
                                kinds: vec![7376],
                                tags: spending_tags,
                                limit: Some(1),
                                relays: self.database.find_relay_candidates(
                                    7376,
                                    &tag_vec[1],
                                    &false,
                                ),
                                cache_first: true,
                                ..Default::default()
                            });
                        }
                    }
                    "e" => {
                        if event_tag.is_none() {
                            event_tag = Some(tag_vec.to_vec());
                            requests.push(Request {
                                ids: vec![tag_vec[1].clone()],
                                kinds: vec![1],
                                relays: self.database.find_relay_candidates(1, "", &false),
                                cache_first: true,
                                ..Default::default()
                            });
                        }
                    }
                    _ => {}
                }
            }
        }

        // Validate essential tags are present
        if proof_tags.is_empty() || mint_tag.is_none() || recipient_tag.is_none() {
            return Err(anyhow!("missing required tags"));
        }

        let mint_url = mint_tag.as_ref().unwrap()[1].clone();
        let recipient = recipient_tag.as_ref().unwrap()[1].clone();

        // Parse nutzap information
        let mut total = 0i32;
        let mut proofs = Vec::new();
        let mut is_p2pk_locked = false;
        let mut p2pk_pubkey: Option<String> = None;

        for proof_tag in proof_tags {
            // Try to extract proof from the tag
            if let Ok(proof_data) = serde_json::from_str::<Value>(&proof_tag[1]) {
                // Basic amount extraction - would need proper cashu types
                if let Some(amount) = proof_data.get("amount").and_then(|a| a.as_i64()) {
                    total += amount as i32;
                }

                // Check for P2PK locking
                if let Some(secret) = proof_data.get("secret").and_then(|s| s.as_str()) {
                    if secret.contains("P2PK") {
                        is_p2pk_locked = true;
                        // Try to extract pubkey from P2PK secret
                        if let Ok(secret_data) = serde_json::from_str::<Value>(secret) {
                            if let Some(array) = secret_data.as_array() {
                                if array.len() >= 2 && array[0].as_str() == Some("P2PK") {
                                    if let Some(data_obj) = array[1].as_object() {
                                        if let Some(pubkey_str) =
                                            data_obj.get("data").and_then(|d| d.as_str())
                                        {
                                            if p2pk_pubkey.is_none() {
                                                p2pk_pubkey = Some(pubkey_str.to_string());
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }

                proofs.push(ProofUnion { data: proof_data });
            }
        }

        let result = Kind9321Parsed {
            amount: total,
            recipient,
            mint_url,
            proofs,
            redeemed: false, // Default to not redeemed, will check later
            comment: if event.content.is_empty() {
                None
            } else {
                Some(event.content.clone())
            },
            is_p2pk_locked,
            p2pk_pubkey,
            event_id: event_tag.map(|tag| tag[1].clone()),
        };

        // Deduplicate requests using the utility
        let deduplicated_requests = RequestDeduplicator::deduplicate_requests(requests);

        Ok((result, Some(deduplicated_requests)))
    }

    pub fn prepare_kind_9321(&self, unsigned_event: &mut UnsignedEvent) -> Result<Event> {
        if unsigned_event.kind.as_u64() != 9321 {
            return Err(anyhow!("event is not kind 9321"));
        }

        // Validate required tags
        let mut has_proof = false;
        let mut has_mint = false;
        let mut has_recipient = false;

        for tag in &unsigned_event.tags {
            let tag_vec = tag.as_vec();
            if tag_vec.len() >= 2 {
                match tag_vec[0].as_str() {
                    "proof" => has_proof = true,
                    "u" => has_mint = true,
                    "p" => has_recipient = true,
                    _ => {}
                }
            }
        }

        if !has_proof {
            return Err(anyhow!("kind 9321 must include at least one proof tag"));
        }

        if !has_mint {
            return Err(anyhow!("kind 9321 must include a u tag with mint URL"));
        }

        if !has_recipient {
            return Err(anyhow!("kind 9321 must include a p tag with recipient"));
        }

        self.signer_manager
            .sign_event(unsigned_event)
            .map_err(|e| anyhow!("failed to sign event: {}", e))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use nostr::{EventBuilder, Keys, Kind, Tag};

    #[test]
    fn test_parse_kind_9321_basic() {
        let keys = Keys::generate();
        let recipient_keys = Keys::generate();
        let mint_url = "https://mint.example.com";
        let proof_json = r#"{"amount":100,"secret":"test_secret","C":"test_C","id":"test_id"}"#;

        let tags = vec![
            Tag::parse(vec!["proof".to_string(), proof_json.to_string()]).unwrap(),
            Tag::parse(vec!["u".to_string(), mint_url.to_string()]).unwrap(),
            Tag::parse(vec![
                "p".to_string(),
                recipient_keys.public_key().to_string(),
            ])
            .unwrap(),
        ];

        let event = EventBuilder::new(Kind::Custom(9321), "Test nutzap", tags)
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let (parsed, requests) = parser.parse_kind_9321(&event).unwrap();

        assert_eq!(parsed.amount, 100);
        assert_eq!(parsed.recipient, recipient_keys.public_key().to_hex());
        assert_eq!(parsed.mint_url, mint_url);
        assert!(!parsed.redeemed);
        assert_eq!(parsed.comment, Some("Test nutzap".to_string()));
        assert!(!parsed.is_p2pk_locked);
        assert!(requests.is_some());
    }

    #[test]
    fn test_parse_kind_9321_missing_required_tags() {
        let keys = Keys::generate();
        let recipient_keys = Keys::generate();

        let tags = vec![
            Tag::parse(vec![
                "p".to_string(),
                recipient_keys.public_key().to_string(),
            ])
            .unwrap(),
            // Missing proof and mint tags
        ];

        let event = EventBuilder::new(Kind::Custom(9321), "Test nutzap", tags)
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let result = parser.parse_kind_9321(&event);

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("missing required tags"));
    }

    #[test]
    fn test_prepare_kind_9321_no_signer() {
        let keys = Keys::generate();
        let recipient_keys = Keys::generate();
        let mint_url = "https://mint.example.com";
        let proof_json = r#"{"amount":100,"secret":"test_secret"}"#;

        let tags = vec![
            Tag::parse(vec!["proof".to_string(), proof_json.to_string()]).unwrap(),
            Tag::parse(vec!["u".to_string(), mint_url.to_string()]).unwrap(),
            Tag::parse(vec![
                "p".to_string(),
                recipient_keys.public_key().to_string(),
            ])
            .unwrap(),
        ];

        let mut event = EventBuilder::new(Kind::Custom(9321), "Test nutzap", tags)
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let result = parser.prepare_kind_9321(&mut event);

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("signing not implemented"));
    }

    #[test]
    fn test_parse_wrong_kind() {
        let keys = Keys::generate();

        let event = EventBuilder::new(Kind::TextNote, "test", Vec::new())
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let result = parser.parse_kind_9321(&event);

        assert!(result.is_err());
    }
}
